app/uk/gov/hmrc/ctcguaranteebalancerouter/connectors/EISConnector.scala

```diff
@@ -93,32 +99,38 @@ class EISConnectorImpl(
             .withBody(request)
             .setHeader(headerCarrier.headersForUrl(headerCarrierConfig)(eisInstanceConfig.accessCodeUrl): _*)
             .execute[Either[UpstreamErrorResponse, HttpResponse]]
-            .map {
-              case Right(httpResponse)         => Right(httpResponse.json)
-              case Left(upstreamErrorResponse) => Left(RoutingError.Upstream(upstreamErrorResponse))
+            .map[Either[ConnectorError, AccessCodeResponse]] {
```

Comment:
Can condense the code a bit by getting execute to deserialise AccessCodeResponse:
```

.execute[Either[UpstreamErrorResponse, AccessCodeResponse]]
            .map {
              case Right(accessCodeResponse) => Right(accessCodeResponse)
              // TODO: Improve once we know what EIS will do
              case Left(UpstreamErrorResponse(_, statusCode, _, _)) if statusCode == BAD_REQUEST || statusCode == NOT_FOUND => Left(ConnectorError.NotFound)
              case Left(upstreamErrorResponse)                                                                              => Left(ConnectorError.Upstream(upstreamErrorResponse))
            }
            .recover {
              case JsResultException(_) => Left(ConnectorError.FailedToDeserialise)
              case NonFatal(e) =>
                logger.error(s"Request Error: Routing to $code failed to retrieve data with message ${e.getMessage}. Correlation ID: $correlationId.")
                Left(ConnectorError.Unexpected("message", Some(e)))
            }

```

Reply:
I feel this is six of one and half a dozen of the other. I decided to do it this way primarily to avoid creating an exception (`JsResultException`) which can be expensive (generating stack traces and all), something we didn't need to do.

Reply:
Fair enough!